streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -462,23 +463,29 @@ public Map<TopicPartition, OffsetAndMetadata> prepareCommit() {
         }
     }
 
-    private Long findOffset(final TopicPartition partition) {
+    private OffsetAndMetadata findOffsetAndMetadata(final TopicPartition partition) {
         Long offset = partitionGroup.headRecordOffset(partition);
+        Optional<Integer> leaderEpoch = partitionGroup.headRecordLeaderEpoch(partition);
+        final long partitionTime = partitionGroup.partitionTimestamp(partition);
         if (offset == null) {
             try {
-                offset = mainConsumer.position(partition);
-            } catch (final TimeoutException error) {
-                // the `consumer.position()` call should never block, because we know that we did process data
-                // for the requested partition and thus the consumer should have a valid local position
-                // that it can return immediately
-
-                // hence, a `TimeoutException` indicates a bug and thus we rethrow it as fatal `IllegalStateException`
-                throw new IllegalStateException(error);
+                if (nextOffsetsAndMetadataToBeConsumed.containsKey(partition)) {
+                    final OffsetAndMetadata offsetAndMetadata = nextOffsetsAndMetadataToBeConsumed.get(partition);
+                    offset = offsetAndMetadata.offset();
+                    leaderEpoch = offsetAndMetadata.leaderEpoch();
+                } else {
+                    // This indicates a bug and thus we rethrow it as fatal `IllegalStateException`
+                    final String errorMsg = String.format("Stream task " + id + " does not know the partition: " + partition);
+                    log.error(errorMsg);
+                    throw new IllegalStateException(errorMsg, new NullPointerException());
```

Comment:
Not sure why you are creating a null pointer exception here?

Reply:
Yes, I also think that the next offset and leader epoch are existing in the map for sure. BTW, I was hesitating to retrieve a key from a map without checking the existence of the key. NullPointer refers to the fact that the key is not existing in the map. But I'm OK to remove it as well.